### ISSUE TITLE
ValueError when using shap_interaction_values

### ISSUE THREAD

**Issue by vruvora** [shap/shap#248]
I am using `LGBM` for my modelling stack and trying to extract out the `shap_interaction_values` and receiving the following error: 

```
ValueError: could not convert string to float: 

'62||160||326||357||856||992||1026||1291||1585||1720||1722||2064||2115||2245||2418||2427||2608||2677||2767||2902||3504||3741||3761||3781||3804||4074'
```

**Comment by slundberg** [comment]
hmm...could you give some kind of minimal way I can reproduce that? Thanks.

**Comment by daniloascione** [comment]
I believe it can be reproduced with LGBM model trained on categorical features. Is this case supported?

**Comment by slundberg** [comment]
Ah. I'll have to debug it when I get a chance. I thought things worked with categorical features, but it has been a while since I wrote that code so I'll have to check.

**Comment by neverfox** [comment]
I can confirm that when using `shap.dependence_plot`, I can't use a categorical variable for the `ind` parameter or run it for a non-categorical if `interaction_index` isn't set to `None` without getting an error related to converting categoricals to floats.

**Comment by neverfox** [comment]
```
import pandas as pd
import shap
import lightgbm as lgb

df = pd.DataFrame({'A': ['Foo', 'Bar', 'Foo', 'Baz'], 'B': [1, 1, 0, 1], 'C': [1, 0, 1, 1]})
df['A'] = df['A'].astype('category')
lgb_train = lgb.Dataset(df[['A', 'B']], df['C'])

gbm = lgb.train({'objective': 'binary', 'min_data': 1, 'min_data_in_bin': 1},
                lgb_train,
                num_boost_round=1)

shap.initjs()
explainer = shap.TreeExplainer(gbm)
shap_values = explainer.shap_values(df[['A', 'B']])

shap.dependence_plot("A", shap_values, df[['A', 'B']])
# ValueError: could not convert string to float: 'Foo'
# Expected plot with categorical values on x-axis

shap.dependence_plot("B", shap_values, df[['A', 'B']])
# ValueError: could not convert string to float: 'Baz'
# Expected plot with categorical values as color separation
```

**Comment by neverfox** [comment]
>  thought things worked with categorical features

I noticed that in [this notebook](http://localhost:8889/notebooks/notebooks/tree_explainer/Census%20income%20classification%20with%20LightGBM.ipynb), categorical features plot fine but that's only because it uses a technique whereby the features are trained as integers but displayed using a mirror of the data that has the category labels (X_display). So it seems that this might have only ever worked if that technique is used and not when the model is trained directly on category dtypes.

**Comment by daniloascione** [comment]
> > thought things worked with categorical features
> 
> I noticed that in [this notebook](http://localhost:8889/notebooks/notebooks/tree_explainer/Census%20income%20classification%20with%20LightGBM.ipynb), categorical features plot fine but that's only because it uses a technique whereby the features are trained as integers but displayed using a mirror of the data that has the category labels (X_display). So it seems that this might have only ever worked if that technique is used and not when the model is trained directly on category dtypes.

Right, I managed to render the plot with a copy of the original dataframe that uses the _codes_ for the categorical features:
`shap.dependence_plot(column, shap_values, df_with_codes, display_features=df_original)`

(LGBM model was trained using a panda dataframe with categorical data)

**Comment by slundberg** [comment]
Thanks for looking into this. @neverfox your comment about it being related to the interaction index is helpful. What is failing seems to be just the fast heuristic for choosing how to color the figure. This heuristic is based on a binned correlation test that of course won't work for categorical data. I think this could be fixed by adding a quick test to see if the current data has categorical columns, and if it does to convert them to ints (essentially doing the `df_with_codes` and `df_original` approach above automatically when needed).

**Comment by DanGolding** [comment]
I have the same issue with a catboost model

**Comment by DanGolding** [comment]
> > thought things worked with categorical features
> 
> I noticed that in [this notebook](http://localhost:8889/notebooks/notebooks/tree_explainer/Census%20income%20classification%20with%20LightGBM.ipynb), categorical features plot fine but that's only because it uses a technique whereby the features are trained as integers but displayed using a mirror of the data that has the category labels (X_display). So it seems that this might have only ever worked if that technique is used and not when the model is trained directly on category dtypes.

@neverfox do you have a public link to the notebook with the mirrored data?

**Comment by neverfox** [comment]
@DanGolding [Here](https://github.com/slundberg/shap/blob/master/notebooks/tree_explainer/Census%20income%20classification%20with%20LightGBM.ipynb)

**Comment by DanGolding** [comment]
@neverfox Thanks!

**Comment by tudstudent** [comment]
Categorical split support across all code paths has been added in #4263, including the interventional path, GPU path, and pure Python pytree fallback.
